# Profitable App Profiles for the App Store and Google Play Markets

* This project is about analysing data sets(about apps) from the Google Play store and the Apple Store. The company only builds apps that are free to download and install and the main source of revenue consists of in-app ads. The revenue is influenced by the number of people who use their apps.

### Goal:
   * The first goal is to analyze data in order to help developers and the company what type of apps are likely to attract more users.
   * The second goal which is more personal, is to help get a better understanding or become more comfortable with the fundamnetals of Python.
   
There are over 2 million apps in the Apple store and 2.1 million in the Google(Android) store. Collecting and analysing that data of 4 million apps requires time and money. That is why we'll work with 2 data sets that were found at no cost.

    1. The data set from Google play we're going to use contains around 10,000 apps. The data was collected in August 2018. You can have access to the data set from [this link](https://www.kaggle.com/lava18/google-play-store-apps)
    
    2. The Apple store data set has around 7000 iOS apps and was collected in July 2017. [Here's](https://www.kaggle.com/ramamet4/app-store-apple-data-set-10k-apps) that'll give you access to the data set.
    
    
We'll start by opening the two data sets to explore. After opening the data sets, we'll isolate the header row from the rest of the data. That'll make it easier for us to come up with accurate result. 

In [1]:
# Opening the Android data set
import csv 
android_set = open('googleplaystore.csv')
read_android_file = csv.reader(android_set)
a_list = list(read_android_file)
android_header = a_list[0] # isolating the header row
android_file = a_list[1:] # rest of the data set without header row

# Opening the ios data set
ios_set = open('AppleStore.csv')
read_ios_file = csv.reader(ios_set)
ios_list = list(read_ios_file)
ios_header = ios_list[0] # isolating the header row
ios_file = ios_list[1:] # rest of the data

We also used the explore_data() below to help us print the rows in a readable way.

In [2]:
# Explore_data function
def explore_data(dataset, start, end, rows_and_columns=False):
# dataset is either ios_file or android_file defined in the 
# previous cell
    dataset_slice = dataset[start:end] # to slice the dataset
    for row in dataset_slice: 
        print(row) # to print the sliced data set
        print('\n') # adds a new (empty) line after each row
        
    if rows_and_columns:
        print('Number of rows:', len(dataset))
        print('Number of columns:', len(dataset[0]))

In [3]:
# printing the Google data set(android_file) & header
print(android_header)
print('\n')
print(explore_data(android_file, 0, 2, True))

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['Coloring book moana', 'ART_AND_DESIGN', '3.9', '967', '14M', '500,000+', 'Free', '0', 'Everyone', 'Art & Design;Pretend Play', 'January 15, 2018', '2.0.0', '4.0.3 and up']


Number of rows: 10841
Number of columns: 13
None


Above we've printed the header and a couple of rows from the Google Store data set. The data has 10841 rows and 13 columns. By looking at the header row, I believe the columns that are going to help in our analysis are: Category, Rating, installs and genres. You can refer to this [documentation](https://www.kaggle.com/lava18/google-play-store-apps) to read about each column's description. 

In [4]:
# printing the Apple store data(ios_file) & header
print(ios_header)
print('\n') 
print(explore_data(ios_file, 0, 2, True))

['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


Number of rows: 7197
Number of columns: 16
None


The Apple data set has 7,197 rows and 16 columns. The prime_genre column will help us in our analysis.
##### Remeber that we excluded the header row from both data sets and they're not included in the count. 

### Data Cleaning
The goal of cleaning our data is to remove duplicates, correct wrong data and modify the data to fit the purpose of our analysis.

The Google play data set has a dedicated [discussion section](https://www.kaggle.com/lava18/google-play-store-apps/discussion), and we can see that [one of the discussions](https://www.kaggle.com/lava18/google-play-store-apps/discussion/66015) describes an error for the row 10472



In [5]:
# Checking row 10472
print(android_header)
print('\n')
print(android_file[10472])
print('\n')
print(android_file[0])

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


['Life Made WI-Fi Touchscreen Photo Frame', '1.9', '19', '3.0M', '1,000+', 'Free', '0', 'Everyone', '', 'February 11, 2018', '1.0.19', '4.0 and up']


['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


The row 10472 correspond to the Life Made WI-FI Photo Frame app. We noticed that the rating '19' is way off because the maximum rating in the Google Store is 5. Actually it's the category that is missing and that causes the column to shift. We could do some research and find this app's category but for now we'll delete it.

In [6]:
print(len(android_file))
del android_file[10472]
print(len(android_file))

10841
10840


### Removing Duplicate entries

If you explore the Google Play data set long enough or look at the discussions section, you'll notice some apps have duplicate entries. Below we notice that there are 1,181 cases where an app occurs more than once in the Google store and 0 in the Apple store.

In [7]:
# checking duplicates for both data set
def duplicate(dataset): # to take any dataset
    duplicate_apps = [] # for duplicates apps
    unique_apps = [] # for unique apps
    for app in dataset:
        name = app[0] # store the name of the app
        if name in unique_apps:
            duplicate_apps.append(name)
        else:
            unique_apps.append(name)
    return duplicate_apps
    return unique_apps
            
android_duplicate = duplicate(android_file)
ios_duplicate = duplicate(ios_file)

print('Number of duplicate adnroid apps:', len(android_duplicate))
print('\n')
print('Examples of duplicate android apps:', android_duplicate[:5])
print('\n')
# checking ios apps
print('Number of duplicate ios apps:', len(ios_duplicate))

Number of duplicate adnroid apps: 1181


Examples of duplicate android apps: ['Quick PDF Scanner + OCR FREE', 'Box', 'Google My Business', 'ZOOM Cloud Meetings', 'join.me - Simple Meetings']


Number of duplicate ios apps: 0


For instance, Instagram has four duplicate entries:

In [8]:
# Printing Instagram for Android
for app in android_file:
    name = app[0]
    if name == 'Instagram':
        print(app)
        print('\n')

['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']


['Instagram', 'SOCIAL', '4.5', '66577446', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']


['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']


['Instagram', 'SOCIAL', '4.5', '66509917', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']




If you examine the rows we printed for the Instagram app above, the main difference happens on the fourth position of each row, which corresponds to the number of reviews (under the column titled 'Reviews'. The different numbers show the data was collected at different times.

We can use this information to build a criterion for removing the duplicates. The higher the number of reviews, the more recent the data should be. Rather than removing duplicates randomly, we'll only keep the row with the highest number of reviews and remove the other entries for any given app, because the higher the number of reviews, the more reliable the ratings.


In [9]:
# create a dictionary where each key is the unique app name
# and the value is highest number of reviews for that app
reviews_max = {}

for app in android_file:
    name = app[0]
    n_reviews = float(app[3])
    if name in reviews_max and reviews_max[name] < n_reviews:
        reviews_max[name] = n_reviews
    elif name not in reviews_max:
        reviews_max[name] = n_reviews
        



Remember previously we found out that there are 1181 duplicate entries in the Google store data set. So the length of our unique apps should be the difference between our dataset and 1181

In [10]:
print('Expected length is:', len(android_file) - 1181)
print('Actual length is:', len(reviews_max))

Expected length is: 9659
Actual length is: 9659


In [11]:
android_clean = [] # our new cleaned data set. Unique apps
already_added = [] # app names

for app in android_file:
    name = app[0]
    n_reviews = float(app[3])
    if (reviews_max[name] == n_reviews) and (name not in already_added):
        android_clean.append(app)
        already_added.append(name)
        


Now let's see and confirm that our new and cleaned data set has 9659 rows.

In [12]:
print(explore_data(android_clean, 0, 3, True))

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


Number of rows: 9659
Number of columns: 13
None


## Removing non-English Apps

#### Part 1

Remember that our company only builds apps for English speaking audience. That's why we have to make sure that our data sets don't include any non-English app. Below are examples of non-English apps from both data sets.

In [13]:
print('From iOS:', ios_file[813][1])
print('\n')
print('From Android:', android_clean[4412][0])

From iOS: 爱奇艺PPS -《欢乐颂2》电视剧热播


From Android: 中国語 AQリスニング


We'll need to remove apps like that and one way to go about it is to remove each app with a name containing a symbo that is not commonyl used in the English text.

English text usually includes letters from the English alphabet, numbers composed of digits from 0 to 9, punctuation marks (., !, ?, ;), and other symbols (+, *, /).

The numbers corresponding to the characters we commonly use in an English text are all in the range 0 to 127, according to the ASCII (American Standard Code for Information Interchange) system. Based on this number range, we can build a function that detects whether a character belongs to the set of common English characters or not. If the number is equal to or less than 127, then the character belongs to the set of common English characters. If an app name contains a character that is greater than 127, then it probably means that the app has a non-English name.

In [14]:
# A function to check if a character ASCII number is greater than
# 127 or not. 
def is_english(a_string):
    for character in a_string:
        if ord(character) > 127:
            return False
        
    return True
    
print(is_english('Instagram'))
print(is_english('爱奇艺PPS -《欢乐颂2》电视剧热播'))
print(is_english('Docs To Go™ Free Office Suite'))
print(is_english('Instachat 😜'))

True
False
False
False


The function seems to work fine, but some English app names use emojis or other symbols (™, — (em dash), – (en dash), etc.) that fall outside of the ASCII range. Because of this, we'll remove useful apps if we use the function in its current form.

For example: The TM symbols and emoji falls way above 127, that is why our is_english() function returned false for 'Docs To Go™ Free Office Suite' and 'Instachat 😜' even though they're English apps.

In [15]:
print(ord('™'))
print(ord('😜'))

8482
128540


If we're going to use the function we've created, we'll lose useful data since many English apps will be incorrectly labeled as non-English. To minimize the impact of data loss, we'll only remove an app if its name has more than three characters with corresponding numbers falling outside the ASCII range.

In [16]:
# Modifying our is_english function
def is_english(a_string):
    non_ascii = 0
    for character in a_string:
        if ord(character) > 127:
            non_ascii += 1 # will increment by 1 everytime
            # there's a non_english character (outside of 127)
            
    if non_ascii > 3: # if there are more than 3 non_english character
        return False
    else:
        return True
    
print(is_english('Docs To Go™ Free Office Suite'))
print(is_english('Instachat 😜'))
print(is_english('爱奇艺PPS -《欢乐颂2》电视剧热播'))

True
True
False


Now we can use the modified is_english() function to filter out non-English apps

In [17]:
android_english = [] # to store English android apps
ios_english = [] # to store English Apple apps

# Check for Android
for app in android_clean:
    name = app[0]
    if is_english(name):
        android_english.append(app)
        
# Check for ios
for app in ios_file:
    name = app[1]
    if is_english(name):
        ios_english.append(app)
        
print(explore_data(android_english, 0, 2, True))
print('\n')
print(explore_data(ios_english, 0, 2, True))

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


Number of rows: 9614
Number of columns: 13
None


['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


Number of rows: 6183
Number of columns: 16
None


We can see that we're left with 9614 Android apps and 6183 iOS apps.

### Isolating Free Apps

Our company only builds free apps and rely on in-app ads for revenue. Both our data sets contain free and non-free apps, so we'll need to isolat only free apps for our analysis. 

In [18]:
android_final = [] # to store our final dataset after all cleaning
ios_final = []

# Android free app
for app in android_english:
    price = app[7]
    if price == '0':
        android_final.append(app)
        
# ios free app
for app in ios_english:
    price = app[4]
    if price == '0.0':
        ios_final.append(app)
        
print(explore_data(android_final, 0, 1, True))
print('\n')
print(explore_data(ios_final, 0, 1, True))

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


Number of rows: 8864
Number of columns: 13
None


['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


Number of rows: 3222
Number of columns: 16
None


Above you can see that we're left with 8864 Android apps and 3222 iOS apps that are free, unique and directed toward English speakers. Now we can start our analysis.

### Most common apps by genre

As you already know, our goal is to determine what kind of apps are likely to attract more users because our revenue is highly influenced by the number of users.

To decrease risk, we developed a validation strategy that include 3 steps:
    
    * Build a minimal Android version of the app, and add it to Google Play
    * If the app has a good response from the users, we develop it further
    * If the app is profitable after six months, we build an ios version of the app and add it to the App store.
    
We need to find apps that are successful in both markets.

Below we explore both dataset and identify the columns we could use to generate frequency tables and find out the most common genre in each market

In [19]:
print(android_header)
print('\n')
print(ios_header)

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


The columns 'category' and 'genres' could help in our analysis for the Google store

Prime_genre and rating_count_tot will help us for the Apple store

Now, below we'll build 2 functions we can use to analyze the frequency tables:
   * One function to generate frequency tables that show percentages
   * Another function we can use to display the percentages in a descending order

In [20]:
# Function to generate frequency table in %
def freq_table(dataset, index):
    table = {}
    total = 0
    
    for row in dataset:
        total += 1
        value = row[index]
        if value in table:
            table[value] += 1
        else:
            table[value] = 1
            
    table_percentages = {}
    for key in table:
        percentage = (table[key] / total) * 100
        table_percentages[key] = percentage
    return table_percentages


# Function to display the % in a descending order       
def display_table(dataset, index):
    table = freq_table(dataset, index)
    table_display = []
    for key in table:
        key_val_as_tuple = (table[key], key)
        table_display.append(key_val_as_tuple)

    table_sorted = sorted(table_display, reverse = True)
    for entry in table_sorted:
        print(entry[1], ':', entry[0])
        

We'll start with the prime_genre column in the Apple Store

In [21]:
print(display_table(ios_final, 11))

Games : 58.16263190564867
Entertainment : 7.883302296710118
Photo & Video : 4.9658597144630665
Education : 3.662321539416512
Social Networking : 3.2898820608317814
Shopping : 2.60707635009311
Utilities : 2.5139664804469275
Sports : 2.1415270018621975
Music : 2.0484171322160147
Health & Fitness : 2.0173805090006205
Productivity : 1.7380509000620732
Lifestyle : 1.5828677839851024
News : 1.3345747982619491
Travel : 1.2414649286157666
Finance : 1.1173184357541899
Weather : 0.8690254500310366
Food & Drink : 0.8069522036002483
Reference : 0.5586592178770949
Business : 0.5276225946617008
Book : 0.4345127250155183
Navigation : 0.186219739292365
Medical : 0.186219739292365
Catalogs : 0.12414649286157665
None


For the Apple store, the general impression is that apps deisgned for fun and distraction purpose are the most common: Gaming apps comes at the top with 58%, followed by entertainment apps with 7%, and then Photo & video apps with 4%. Then we noticed practical purposes apps such as educational apps comes at 4th with 3.6%. 

Fun apps make up more than half of the Apple store data set but that doesn't necessarly mean that there's a high demand for fun apps. 

In [22]:
# Analyzing categories in Google store
print(display_table(android_final, 1))

FAMILY : 18.907942238267147
GAME : 9.724729241877256
TOOLS : 8.461191335740072
BUSINESS : 4.591606498194946
LIFESTYLE : 3.9034296028880866
PRODUCTIVITY : 3.892148014440433
FINANCE : 3.7003610108303246
MEDICAL : 3.531137184115524
SPORTS : 3.395758122743682
PERSONALIZATION : 3.3167870036101084
COMMUNICATION : 3.2378158844765346
HEALTH_AND_FITNESS : 3.0798736462093865
PHOTOGRAPHY : 2.944494584837545
NEWS_AND_MAGAZINES : 2.7978339350180503
SOCIAL : 2.6624548736462095
TRAVEL_AND_LOCAL : 2.33528880866426
SHOPPING : 2.2450361010830324
BOOKS_AND_REFERENCE : 2.1435018050541514
DATING : 1.861462093862816
VIDEO_PLAYERS : 1.7937725631768955
MAPS_AND_NAVIGATION : 1.3989169675090252
FOOD_AND_DRINK : 1.2409747292418771
EDUCATION : 1.1620036101083033
ENTERTAINMENT : 0.9589350180505415
LIBRARIES_AND_DEMO : 0.9363718411552346
AUTO_AND_VEHICLES : 0.9250902527075812
HOUSE_AND_HOME : 0.8235559566787004
WEATHER : 0.8009927797833934
EVENTS : 0.7107400722021661
PARENTING : 0.6543321299638989
ART_AND_DESIGN : 

For Google store, apps in the category family are the most common with 18. 9%, followed by gaming apps and then pratical purposes apps such as tools, business, lifestyle etc. However, after some investigation, we found that apps in the family category are mostly games for kids.

In [23]:
# few examples of apps in the Google store family category
for row in android_final:
    category = row[1]
    if category == 'FAMILY':
        print(row[0], ':', row[1])

Jewels Crush- Match 3 Puzzle : FAMILY
Coloring & Learn : FAMILY
Mahjong : FAMILY
Super ABC! Learning games for kids! Preschool apps : FAMILY
Toy Pop Cubes : FAMILY
Educational Games 4 Kids : FAMILY
Candy Pop Story : FAMILY
Princess Coloring Book : FAMILY
Hello Kitty Nail Salon : FAMILY
Candy Smash : FAMILY
Happy Fruits Bomb - Cube Blast : FAMILY
Princess Adventures Puzzles : FAMILY
Kids Educational Game 3 Free : FAMILY
Puzzle Kids - Animals Shapes and Jigsaw Puzzles : FAMILY
Coloring book moana : FAMILY
Baby Panda Care : FAMILY
Kids Educational :All in One : FAMILY
Number Counting games for toddler preschool kids : FAMILY
Learn To Draw Glow Flower : FAMILY
No. Color - Color by Number, Number Coloring : FAMILY
Draw.ly - Color by Number Pixel Art Coloring : FAMILY
Baby puzzles : FAMILY
Garden Fruit Legend : FAMILY
Barbie™ Fashion Closet : FAMILY
Candy Day : FAMILY
Learn To Draw Glow Princess : FAMILY
ABC Kids - Tracing & Phonics : FAMILY
Barbie Magical Fashion : FAMILY
Minion Rush: Despi

Kids game such the ones listed below are included in the Family category in the Google store. Side note: these games sounds fun, no wonder kids today are not going playing outside anymore lol.

Jewels Crush- Match 3 Puzzle : FAMILY
Candy Smash : FAMILY
Kids Educational Game 3 Free : FAMILY
Mahjong : FAMILY
Happy Fruits Bomb - Cube Blast : FAMILY

Now let's look at the genre in Google store

In [24]:
# display genre
print(display_table(android_final, 9))

Tools : 8.449909747292418
Entertainment : 6.069494584837545
Education : 5.347472924187725
Business : 4.591606498194946
Productivity : 3.892148014440433
Lifestyle : 3.892148014440433
Finance : 3.7003610108303246
Medical : 3.531137184115524
Sports : 3.463447653429603
Personalization : 3.3167870036101084
Communication : 3.2378158844765346
Action : 3.1024368231046933
Health & Fitness : 3.0798736462093865
Photography : 2.944494584837545
News & Magazines : 2.7978339350180503
Social : 2.6624548736462095
Travel & Local : 2.3240072202166067
Shopping : 2.2450361010830324
Books & Reference : 2.1435018050541514
Simulation : 2.0419675090252705
Dating : 1.861462093862816
Arcade : 1.8501805054151623
Video Players & Editors : 1.7712093862815883
Casual : 1.7599277978339352
Maps & Navigation : 1.3989169675090252
Food & Drink : 1.2409747292418771
Puzzle : 1.128158844765343
Racing : 0.9927797833935018
Role Playing : 0.9363718411552346
Libraries & Demo : 0.9363718411552346
Auto & Vehicles : 0.9250902527075

We notice that the category and genre columns are not that much different. They're both topped by fun and pratical purpose apps. 

The Genres column has more categories. We're only looking for the bigger picture at the moment, so we'll only work with the Category column moving forward.

So far, we found that most of the apps in the Apple store are designed for fun and entertainment purpose, while the Google store is more balanced between apps designed for fun and practical purpose

### Most popular apps by Genre in the App Store

Now we want to get an idea of the kind of apps with the most users. We'll calculate the the average number of installs for each app genre. For the Google Play data set, we can find this information in the Installs column, but this information is missing for the App Store data set. As a workaround, we'll take the total number of user ratings as a proxy, which we can find in the rating_count_tot app.

We'll start with Apple store

In [25]:
# Prime_genre frequecy table
genre_ios = freq_table(ios_final, 11)

for genre in genre_ios:
    total = 0 # sum of user ratings
    len_genre = 0 # number of apps specific to each genre
    for app in ios_final:
        genre_app = app[11]
        if genre_app == genre:
            n_ratings = float(app[5])
            total += n_ratings
            len_genre += 1
    avg_n_ratings = total / len_genre
    print(genre, ':', avg_n_ratings)

Shopping : 26919.690476190477
Catalogs : 4004.0
Utilities : 18684.456790123455
Travel : 28243.8
News : 21248.023255813954
Food & Drink : 33333.92307692308
Weather : 52279.892857142855
Music : 57326.530303030304
Games : 22788.6696905016
Navigation : 86090.33333333333
Social Networking : 71548.34905660378
Book : 39758.5
Finance : 31467.944444444445
Lifestyle : 16485.764705882353
Business : 7491.117647058823
Sports : 23008.898550724636
Education : 7003.983050847458
Health & Fitness : 23298.015384615384
Entertainment : 14029.830708661417
Productivity : 21028.410714285714
Reference : 74942.11111111111
Medical : 612.0
Photo & Video : 28441.54375


We notice that navigation apps have the most user reviews on average, followed by reference app, then social networking, and music apps. 

Let's explore a little deeper and find out about Navigation and Reference genre.

In [26]:
# Navigation apps in the App store
for row in ios_final:
    if row[11] == 'Navigation':
        print(row[1], ':', row[5])

Waze - GPS Navigation, Maps & Real-time Traffic : 345046
Google Maps - Navigation & Transit : 154911
Geocaching® : 12811
CoPilot GPS – Car Navigation & Offline Maps : 3582
ImmobilienScout24: Real Estate Search in Germany : 187
Railway Route Search : 5


We can see that the navigation apps have a lot of users reviews because they include Waze and Google Maps, two well known apps. Those two make up over 500,000 users reviews. 

In [27]:
for row in ios_final:
    if row[11] == 'Reference':
        print(row[1], ':', row[5])

Bible : 985920
Dictionary.com Dictionary & Thesaurus : 200047
Dictionary.com Dictionary & Thesaurus for iPad : 54175
Google Translate : 26786
Muslim Pro: Ramadan 2017 Prayer Times, Azan, Quran : 18418
New Furniture Mods - Pocket Wiki & Game Tools for Minecraft PC Edition : 17588
Merriam-Webster Dictionary : 16849
Night Sky : 12122
City Maps for Minecraft PE - The Best Maps for Minecraft Pocket Edition (MCPE) : 8535
LUCKY BLOCK MOD ™ for Minecraft PC Edition - The Best Pocket Wiki & Mods Installer Tools : 4693
GUNS MODS for Minecraft PC Edition - Mods Tools : 1497
Guides for Pokémon GO - Pokemon GO News and Cheats : 826
WWDC : 762
Horror Maps for Minecraft PE - Download The Scariest Maps for Minecraft Pocket Edition (MCPE) Free : 718
VPN Express : 14
Real Bike Traffic Rider Virtual Reality Glasses : 8
教えて!goo : 0
Jishokun-Japanese English Dictionary & Translator : 0


Again we see that the average user reviews is skewed by the Bible and Dictionary app, both of them are well known and have a lot of users. This pattern continues even with the music and social networking apps because they include popular apps such as Facebook, Spotify, Shazam, Pinterest and so on. 

However, we could explore the reference genre because it has some potential. One thing we could do is take another popular book and turn it into an app where we could add different features besides the raw version of the book. This might include daily quotes from the book, an audio version of the book, quizzes about the book, etc. On top of that, we could also embed a dictionary within the app, so users don't need to exit our app to look up words in an external app.

This idea seems to fit well with the fact that the App Store is dominated by for-fun apps. This suggests the market might be a bit saturated with for-fun apps, which means a practical app might have more of a chance to stand out among the huge number of apps on the App Store.

### Most Popular apps by Genre on Google Play

We have the data about the number of app installs for Google Play. However the install numbers are not precise enough. Down below we can see that the values are open-ended: 100+, 1000+, etc

In [28]:
display_table(android_final, 5)

1,000,000+ : 15.726534296028879
100,000+ : 11.552346570397113
10,000,000+ : 10.548285198555957
10,000+ : 10.198555956678701
1,000+ : 8.393501805054152
100+ : 6.915613718411552
5,000,000+ : 6.825361010830325
500,000+ : 5.561823104693141
50,000+ : 4.7721119133574
5,000+ : 4.512635379061372
10+ : 3.5424187725631766
500+ : 3.2490974729241873
50,000,000+ : 2.3014440433213
100,000,000+ : 2.1322202166064983
50+ : 1.917870036101083
5+ : 0.78971119133574
1+ : 0.5076714801444043
500,000,000+ : 0.2707581227436823
1,000,000,000+ : 0.22563176895306858
0+ : 0.04512635379061372
0 : 0.01128158844765343


For instance, we don't know whether an app with 100,000+ installs has 100,000 installs, 200,000, or 350,000. However, we don't need very precise data for our purposes — we only want to get an idea which app genres attract the most users, and we don't need perfect precision with respect to the number of users.

We're going to leave the numbers as they are, which means that we'll consider that an app with 100,000+ installs has 100,000 installs, and an app with 1,000,000+ installs has 1,000,000 installs, and so on.

To perform computations, however, we'll need to convert each install number to float — this means that we need to remove the commas and the plus characters, otherwise the conversion will fail and raise an error. We'll do this directly in the loop below, where we also compute the average number of installs for each genre (category).

In [29]:
android_category = freq_table(android_final, 1)

for category in android_category:
    total = 0
    len_category = 0
    for app in android_final:
        category_app = app[1]
        if category_app == category:
            n_installs = app[5]
            n_installs = n_installs.replace('+', '')
            n_installs = n_installs.replace(',', '')
            total += float(n_installs)
            len_category += 1
    avg_installs = total / len_category
    print(category, ':', avg_installs)

PERSONALIZATION : 5201482.6122448975
GAME : 15588015.603248259
PHOTOGRAPHY : 17840110.40229885
ART_AND_DESIGN : 1986335.0877192982
SPORTS : 3638640.1428571427
VIDEO_PLAYERS : 24727872.452830188
ENTERTAINMENT : 11640705.88235294
EVENTS : 253542.22222222222
HEALTH_AND_FITNESS : 4188821.9853479853
HOUSE_AND_HOME : 1331540.5616438356
FINANCE : 1387692.475609756
BOOKS_AND_REFERENCE : 8767811.894736841
DATING : 854028.8303030303
LIFESTYLE : 1437816.2687861272
BUSINESS : 1712290.1474201474
COMICS : 817657.2727272727
PARENTING : 542603.6206896552
MEDICAL : 120550.61980830671
FAMILY : 3695641.8198090694
BEAUTY : 513151.88679245283
WEATHER : 5074486.197183099
NEWS_AND_MAGAZINES : 9549178.467741935
EDUCATION : 1833495.145631068
FOOD_AND_DRINK : 1924897.7363636363
SOCIAL : 23253652.127118643
PRODUCTIVITY : 16787331.344927534
AUTO_AND_VEHICLES : 647317.8170731707
LIBRARIES_AND_DEMO : 638503.734939759
SHOPPING : 7036877.311557789
TOOLS : 10801391.298666667
TRAVEL_AND_LOCAL : 13984077.710144928
COMMU

Communication apps seems to have to most installs. Keep in mind that there apps in that category with billion of users: WhatsApp, Facebook Messenger, Gmail, etc). The number is heavily skewed by these apps. 

In [30]:
# Google play communication apps
for app in android_final:
    if app[1] == 'COMMUNICATION' and (app[5] == '1,000,000,000+' 
                                     or app[5] == '500,000,000+'
                                     or app[5] == '100,000,000+'):
        print(app[0], ':', app[5])

WhatsApp Messenger : 1,000,000,000+
imo beta free calls and text : 100,000,000+
Android Messages : 100,000,000+
Google Duo - High Quality Video Calls : 500,000,000+
Messenger – Text and Video Chat for Free : 1,000,000,000+
imo free video calls and chat : 500,000,000+
Skype - free IM & video calls : 1,000,000,000+
Who : 100,000,000+
GO SMS Pro - Messenger, Free Themes, Emoji : 100,000,000+
LINE: Free Calls & Messages : 500,000,000+
Google Chrome: Fast & Secure : 1,000,000,000+
Firefox Browser fast & private : 100,000,000+
UC Browser - Fast Download Private & Secure : 500,000,000+
Gmail : 1,000,000,000+
Hangouts : 1,000,000,000+
Messenger Lite: Free Calls & Messages : 100,000,000+
Kik : 100,000,000+
KakaoTalk: Free Calls & Text : 100,000,000+
Opera Mini - fast web browser : 100,000,000+
Opera Browser: Fast and Secure : 100,000,000+
Telegram : 100,000,000+
Truecaller: Caller ID, SMS spam blocking & Dialer : 100,000,000+
UC Browser Mini -Tiny Fast Private & Secure : 100,000,000+
Viber Mess

The numbers will significantly go down if we remove the apps above. 

In [31]:
# removing 100M+ install apps
under_100m = []
for app in android_final:
    n_installs = app[5]
    n_installs = n_installs.replace('+', '')
    n_installs = n_installs.replace(',', '')
    if app[1] == 'COMMUNICATION' and (float(n_installs) < 100000000):
        under_100m.append(float(n_installs))
     
print('Average install of apps under 100M:', sum(under_100m) / len(under_100m)) 

Average install of apps under 100M: 3603485.3884615386


We've also noticed the same pattern in category such as productivity (includes apps from Microsft), video players (Youtube) and so one. The categories are saturated with apps and it'll be hard to compete because there are giants like Google, Facebook and so on. 

Remember our aim is to find and recommend a niche/app that would be ideal for Android and iOS. Previously, when we did the Apple store analysis, we ended up recommending developing an app in the 'reference' genre. We could explore the same genre in Google store as well.

In [32]:
for app in android_final:
    if app[1] == 'BOOKS_AND_REFERENCE':
        print(app[0], ':', app[5])

E-Book Read - Read Book for free : 50,000+
Download free book with green book : 100,000+
Wikipedia : 10,000,000+
Cool Reader : 10,000,000+
Free Panda Radio Music : 100,000+
Book store : 1,000,000+
FBReader: Favorite Book Reader : 10,000,000+
English Grammar Complete Handbook : 500,000+
Free Books - Spirit Fanfiction and Stories : 1,000,000+
Google Play Books : 1,000,000,000+
AlReader -any text book reader : 5,000,000+
Offline English Dictionary : 100,000+
Offline: English to Tagalog Dictionary : 500,000+
FamilySearch Tree : 1,000,000+
Cloud of Books : 1,000,000+
Recipes of Prophetic Medicine for free : 500,000+
ReadEra – free ebook reader : 1,000,000+
Anonymous caller detection : 10,000+
Ebook Reader : 5,000,000+
Litnet - E-books : 100,000+
Read books online : 5,000,000+
English to Urdu Dictionary : 500,000+
eBoox: book reader fb2 epub zip : 1,000,000+
English Persian Dictionary : 500,000+
Flybook : 500,000+
All Maths Formulas : 1,000,000+
Ancestry : 5,000,000+
HTC Help : 10,000,000+
E

The book and reference category includes a few popular apps with over 100M installs. Let's look at them below

In [35]:
# Poppular book apps
for app in android_final:
    if app[1] == 'BOOKS_AND_REFERENCE' and (app[5] == '1,000,000,000+'
                                            or app[5] == '500,000,000+'
                                            or app[5] == '100,000,000+'):
        print(app[0], ':', app[5])

Google Play Books : 1,000,000,000+
Bible : 100,000,000+
Amazon Kindle : 100,000,000+
Wattpad 📖 Free Books : 100,000,000+
Audiobooks from Audible : 100,000,000+


Even though there are popular apps in this category, there is still potential. This category is alread full of apps for reading ebooks, libraries, dictionaries, translators and so on. 

We could combine social networking with ebooks. Develop an app around a popular book with features such as: group or private chat, friend request, etc

## Conclusion

In this project, we analyzed data from the Google Plaay store and the Apple Store. The data sets contained data about applications different applications from both platform. The goal was to recommend an app profile that can be successful on Android and iOS. 

We concluded that developing an app about a popular book with social networking features can be profitable. 